### PR TITLE
Close #60: Multisample FBO support

### LINKED ISSUE #60: Implement planes_samples attribute for FBO windows
Creating a multisample FBO requires use of the appropriate extension and blitting into a normal FBO before readback.

### PR DESCRIPTION
**Pull request by tribal-tec**

### CODE REVIEW

File: eq/client/glWindow.cpp [hunk 1]
```diff
@@ -139,44 +142,63 @@ bool GLWindow::configInitFBO()
...
     }
 
-    // needs glew initialized (see above)
-    _impl->fbo = new util::FrameBufferObject( &_impl->glewContext );
+    if( !_initFBO( _impl->fbo, 0 ))
+        return false;
+
+    int samplesSize = getIAttribute( WindowSettings::IATTR_PLANES_SAMPLES );
+    if( samplesSize == AUTO )
+        return true;
```
Comment (eile) [inline eq/client/glWindow.cpp]: what if samplesSize in <= 0?

File: eq/client/glWindow.cpp [hunk 2]
```diff
@@ -139,44 +142,63 @@ bool GLWindow::configInitFBO()
...
+    return _initFBO( _impl->fboMultiSample, samplesSize );
+}
+
+void GLWindow::configExitFBO()
+{
+    _exitFBO( _impl->fboMultiSample );
+    _exitFBO( _impl->fbo );
+}
 
+bool GLWindow::_initFBO( util::FrameBufferObject*& fbo, const int samplesSize )
```
Comment (eile) [inline eq/client/glWindow.cpp]: Imo _createFBO/_destroyFBO is more in line with the other system window impls.

**Comment by tribal-tec** [comment]
Updated

**Comment by tribal-tec** [comment]
retest this please

File: eq/client/systemWindow.h [hunk 3]
```diff
@@ -77,6 +77,12 @@ class SystemWindow
     /** Bind the window's FBO, if it uses an FBO drawable. @version 1.0 */
     EQ_API virtual void bindFrameBuffer() const = 0;
 
+    /** Bind the window's draw FBO, used for multisampling. @version 1.9 */
+    EQ_API virtual void bindDrawFrameBuffer() const = 0;
+
+    /** Update the window's FBO from the multisampled FBO. @version 1.9 */
+    EQ_API virtual void updateFrameBuffer() const = 0;
```
Comment (eile) [inline eq/client/systemWindow.h]: updateDrawFrameBuffer?

File: eq/fabric/error.cpp [hunk 4]
```diff
@@ -54,7 +55,7 @@ ErrorData _errors[] = {
       "Framebuffer incomplete, missing read buffer" },
     { ERROR_FRAMEBUFFER_INITIALIZED, "FBO already initialized" },
     { ERROR_FRAMEBUFFER_INVALID_SIZE, "FBO size not supported" },
-
+    { ERROR_FRAMEBUFFER_INVALID_SAMPLES, "FBO multisamples not supported" },
```
Comment (eile) [inline eq/fabric/error.cpp]: Multisampled FBO not supported

File: eq/util/texture.cpp [hunk 5]
```diff
@@ -304,11 +305,18 @@ void Texture::bindToFBO( const GLenum target, const int32_t width,
...
 
     EQ_GL_CALL( glBindTexture( _impl->target, _impl->name ));
-    EQ_GL_CALL( glTexImage2D( _impl->target, 0, _impl->internalFormat, width, height, 0,
-                              _impl->format, _impl->type, 0 ));
+    EQ_GL_CALL( glTexImage2D( _impl->target, 0, _impl->internalFormat, width,
+                              height, 0, _impl->format, _impl->type, 0 ));
     EQ_GL_CALL( glFramebufferTexture2DEXT( GL_FRAMEBUFFER, target,
                                            _impl->target, _impl->name, 0 ));
 
+    if( samples > 1 )
```
Comment (eile) [inline eq/util/texture.cpp]: >= 1 or > 0 ?

Comment (eile) [inline eq/util/texture.cpp]: Ah, got it. One sample = no multisampling. eq::ON is 1, is this translated to non-1 somewhere?

File: eq/client/systemWindow.h (hunk 3 above)
Comment (tribal-tec) [inline eq/client/systemWindow.h]: no. If at all, it's updateRenderFrameBuffer(), the FBO used by bindFrameBuffer().

Comment (eile) [inline eq/client/systemWindow.h]: True.

File: eq/util/texture.cpp (hunk 5 above)
Comment (tribal-tec) [inline eq/util/texture.cpp]: ? Can't follow here.

**Comment by tribal-tec** [comment]
retest this please

File: eq/util/texture.cpp (hunk 5 above)
Comment (eile) [inline eq/util/texture.cpp]: Nevermind

**Comment by eile** [comment]
+2

**Pull request merged**